Mathlib/FieldTheory/Perfect.lean

```diff
@@ -0,0 +1,104 @@
+   separable.
+
+-/
+
+open Function
+
+/-- A perfect ring of characteristic `p` (prime) in the sense of Serre.
+
+NB: This is not related to the concept with the same name introduced by Bass (related to projective
+covers of modules). -/
+class PerfectRing (R : Type _) (p : ℕ) [CommRing R] [Fact p.Prime] [CharP R p] : Prop where
+  bijective_frobenius : Bijective $ frobenius R p
+
+section PerfectRing
+
+variable (R : Type _) (p : ℕ) [CommRing R] [Fact p.Prime] [CharP R p]
+
+/-- For a reduced ring, surjectivity of the Frobenius map is a sufficient condition for perfection.
+-/
+lemma mkOfReduced [IsReduced R] (h : Surjective $ frobenius R p) : PerfectRing R p :=
+  ⟨frobenius_inj R p, h⟩
+
+instance ofFiniteOfIsReduced [Finite R] [IsReduced R] : PerfectRing R p :=
+  mkOfReduced _ _ $ Finite.surjective_of_injective (frobenius_inj R p)
+
+variable [PerfectRing R p]
+
+@[simps!]
+noncomputable def frobeniusEquiv : R ≃+* R :=
+  RingEquiv.ofBijective (frobenius R p) PerfectRing.bijective_frobenius
+#align frobenius_equiv frobeniusEquiv
+
+@[simp]
+theorem coe_frobeniusEquiv : ⇑(frobeniusEquiv R p) = frobenius R p := rfl
+#align coe_frobenius_equiv coe_frobeniusEquiv
+
+@[simp]
+theorem frobenius_comp_frobeniusEquiv_symm :
+    (frobenius R p).comp (frobeniusEquiv R p).symm = RingHom.id R := by
+  ext; simp [surjInv_eq]
+
+lemma polynomial_expand_eq (f : Polynomial R) :
+    Polynomial.expand R p f = (f.map (frobeniusEquiv R p).symm) ^ p := by
+  rw [← (f.map (S := R) (frobeniusEquiv R p).symm).expand_char p, Polynomial.map_expand,
+    Polynomial.map_map, frobenius_comp_frobeniusEquiv_symm, Polynomial.map_id]
+
+@[simp]
+theorem not_irreducible_expand (f : Polynomial R) : ¬ Irreducible (Polynomial.expand R p f) := by
+  have hp : Fact p.Prime := inferInstance
+  rw [polynomial_expand_eq]
+  exact fun hf ↦ hf.not_unit $ (of_irreducible_pow hp.out.ne_one hf).pow p
+
+end PerfectRing
+
+/-- A perfect field.
+
+See also `PerfectRing` for a generalisation in positive characteristic. -/
+class PerfectField (K : Type _) [Field K] : Prop where
+  separable_of_irreducible : ∀ {f : Polynomial K}, Irreducible f → f.Separable
+
+lemma PerfectRing.toPerfectField (K : Type _) (p : ℕ)
+    [Field K] [hp : Fact p.Prime] [CharP K p] [PerfectRing K p] : PerfectField K := by
+  refine' PerfectField.mk $ fun {f} hf ↦ _
+  rcases Polynomial.separable_or p hf with h | ⟨-, g, -, rfl⟩
+  · assumption
+  · exfalso; revert hf; simp
+
+namespace PerfectField
+
+variable (K : Type _) [Field K]
+
+instance ofCharZero [CharZero K] : PerfectField K := ⟨Irreducible.separable⟩
+
+instance ofFinite [Finite K] : PerfectField K := by
+  obtain ⟨p, _instP⟩ := CharP.exists K
+  have : Fact p.Prime := ⟨CharP.char_is_prime K p⟩
+  exact PerfectRing.toPerfectField K p
+
+variable [PerfectField K]
+
+instance PerfectField.toPerfectRing (p : ℕ) [Fact p.Prime] [CharP K p] : PerfectRing K p := by
```

Comment:
I have a half-baked attempt using `Mathlib.FieldTheory.SplittingField.Construction` with a lot of `sorry`s:
```lean
instance PerfectField.toPerfectRing (p : ℕ) [HP : Fact p.Prime] [CharP K p] : PerfectRing K p := by
  refine' mkOfReduced _ _ $ fun y ↦ _
  simp only [frobenius, RingHom.coe_mk, MonoidHom.coe_mk, OneHom.coe_mk]
  let f : Polynomial K := Polynomial.X ^ p - Polynomial.C y
  have hfdeg : f.degree = p :=
    Polynomial.degree_X_pow_sub_C (Nat.pos_of_ne_zero <| NeZero.ne p) y
  have hfdeg' : f.degree ≠ 0 := by rw [hfdeg]; norm_cast; exact NeZero.ne p
  have hfnotunit : ¬ IsUnit f := fun h => hfdeg' <| Polynomial.degree_eq_zero_of_isUnit h
  have hf' : Polynomial.derivative f = 0 := by
    rw [map_sub, Polynomial.derivative_X_pow, CharP.cast_eq_zero, map_zero, zero_mul,
      Polynomial.derivative_C, sub_self]
  have hfnotirr : ¬ Irreducible f := fun h => by
    replace h := separable_of_irreducible h
    rw [Polynomial.Separable, hf', isCoprime_zero_right] at h
    exact hfnotunit h
  let g := f.factor
  let d := g.natDegree
  have hd0 : 0 < d := by
    sorry
  have hdp : d < p := by
    sorry
  have hdinv : IsUnit (d : K) := by
    sorry
  have : Fact (Irreducible g) := f.fact_irreducible_factor
  have hdiv : g ∣ f := Polynomial.factor_dvd_of_not_isUnit hfnotunit
  have hroot := hdiv
  rw [← AdjoinRoot.mk_eq_zero, map_sub, map_pow, AdjoinRoot.mk_X, AdjoinRoot.mk_C,
    sub_eq_zero] at hroot
  have : CharP (AdjoinRoot g) p := by
    sorry
  have hf1 : f.map (AdjoinRoot.of g)
      = ((Polynomial.X (R := AdjoinRoot g)) - Polynomial.C (AdjoinRoot.root g)) ^ p := by
    rw [Polynomial.map_sub, Polynomial.map_pow, Polynomial.map_X,
      Polynomial.map_C, ← hroot, map_pow, sub_pow_char]
  have hg1 : g.map (AdjoinRoot.of g)
      = ((Polynomial.X (R := AdjoinRoot g)) - Polynomial.C (AdjoinRoot.root g)) ^ d := by
    sorry
  have hcoeff : (g.map (AdjoinRoot.of g)).coeff (d - 1) = -d * AdjoinRoot.root g := by
    sorry
  rw [Polynomial.coeff_map] at hcoeff
  use -(d : K)⁻¹ * g.coeff (d - 1)
  sorry
```